moby/moby
# Upgrade to aws go sdk v2 for cloudwatch logging driver

**- What I did**
Go SDK v1 is pretty outdated. In particular, it struggles with things like error handling and rate limits. I upgraded the awslogs logging driver to use the v2 SDK.

**- How I did it**
Bumped the imports, made a bunch of code mods, verified it works

**- How to verify it**
I verified in 2 ways.

Firstly, tests. Note that credential tests are very hard to verify now. It seems SDKv2 doesn't have a good way to tell your client to _only_ get creds. Also though, we mainly piggyback on aws libs to get creds, so I'm not sure we need to test this.

**I'm not confident this won't break someone. But I don't see a great way to test. I'm hoping someone can find a good way to validate that we haven't broken functionality.**

I also _changed_ the behaviour of the user agent tagging. I'm not sure this matters. We'll probably ruin some product-manager's day at AWS. But now we're using their blessed framework to add UA, so it might be okay?

Other than that, tests all pass with minor modifications (and no changes in behaviour to my knowledge).

I then went ahead and compiled dockerd and threw this on an aws box. Can confirm, for basic logging, this works great:
```
[ec2-user@ip-10-0-103-225 ~]$ docker run -it --rm --log-driver awslogs --log-opt awslogs-group=joshtest1 ubuntu:22.04 echo josh-what-123
```

I added some logging by dockerd in testing:
```
Sep 03 01:13:23 ip-10-0-103-225.us-west-2.compute.internal dockerd[4862]: time="2022-09-03T01:13:23.859298821Z" level=info msg=JOSH-CREATING
```

So clearly it called into my build. And then we see the logs in real AWS:
![Screenshot 2022-09-03 at 01-14-11 CloudWatch Management Console](https://user-images.githubusercontent.com/6404517/188250829-91e140f8-51aa-4545-8f28-a4b383a15c3c.png)

So I have some confidence that this change doesn't catastrophically break things.

**- Description for the changelog**
Upgrade awslogs loggging aws sdk version to v2


**- A picture of a cute animal (not mandatory but encouraged)**
![](https://hagerty-media-prod.imgix.net/2020/06/1989-Taurus-SHO-Ford-Media-e1591901197302.jpg?auto=format%2Ccompress&ixlib=php-3.3.0)
